### PR TITLE
Move getPointReaderOrNull into AggregatorBase

### LINKED ISSUE #55552: ValuesSource support for reading points
There are a few aggregations where we could really use some kind of clean way to get access to the "points" because we can iterate those in sorted order for the whole shard. Specifically `composite`,  `min`, `max`, `date_histogram` and `auto_date_histogram` look at the points. We'd love to have some nice way for them to do it that tells folks when they implement a new type that this is something that they should think about.

`min` and `max` look up the minimum or maximum value and cast it to a double. `date_histogram` and, soon, `auto_date_histogram` look up the minimum and maximum dates so they can deal with rounding more efficiently. `composite` iterates values is sorted order.

### PR DESCRIPTION
**Pull request by not-napoleon**
This is a pretty straight forward refactoring to move the point reader optimization that min and max use up to the aggregator base, and move the values source config related logic into values source config.  Hopefully that makes it easier to reuse, and also puts the check for if it can be applied in a more visible place.

Closes #55552

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java [hunk 1]
```diff
@@ -105,6 +108,22 @@ public ScoreMode scoreMode() {
         addRequestCircuitBreakerBytes(DEFAULT_WEIGHT);
     }
 
+    /**
+     * Returns a converter for point values if early termination is applicable to
+     * the context or <code>null</code> otherwise.
+     *
+     * @param config The config for the values source metric.
+     */
+    public Function<byte[], Number> getPointReaderOrNull(ValuesSourceConfig config) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java:117]: Do you think now'd be a good time to change the name? I always thought something like `pointReaderIfSafe` or something is a little more clear about *why* you might get null back.

And it isn't so much "if early termination is applicable" as "if there are no filters and the underlying index is in the same order as the values produced by the config"

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java:117]: Could you make the method `final`?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 2]
```diff
@@ -371,4 +374,24 @@ public ValuesSource getValuesSource() {
     public boolean hasGlobalOrdinals() {
         return valuesSource.hasGlobalOrdinals();
     }
+
+    @Nullable
+    public Function<byte[], Number> getPointReaderOrNull() {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:379]: It's probably worth some javadoc on this one too.

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java (hunk 1 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java:117]: > the underlying index is in the same order as the values produced by the config

Out of curiosity, what in the current logic looks at index ordering?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java [hunk 3]
```diff
@@ -85,6 +87,15 @@
...
                                 LongSupplier nowSupplier);
 
+    /**
+     * Attempts to return a reader function for the indexed data of this field.  Some aggregations are able to use this as an optimization
+     * instead of relying on doc values, when the index sort order matches that of the aggregation.
+     *
+     * @param fieldType The field type we want to get a reader for
+     * @return null if we can't get a reader (e.g. because the field is the wrong type), otherwise a point reader function.
+     */
+    default Function<byte[], Number> getPointReader(MappedFieldType fieldType) { return null; }
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java:97]: I'm not convinced this is the right place to add extensibility here.  The other option I'm considering, we could make `parsePoint` a method on `MappedFieldType` (or, maybe a method that returns a reference to `parsePoint` or null, depending on if it's implemented for that field).  That would let us get rid of the instance type checking, and provide a way for plugins that add fields which use CoreVSTs to add custom point readers.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java:97]: I agree that it should be the `MappedFieldType`'s responsibility. It is built by the `FieldMapper` which is ultimately the only thing that actually knows how the index was built.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Comment by nik9000** [comment]
@not-napoleon I edited your description to say that this closes #58769 which I believe it does, in that it creates the abstraction that other aggs can use to read the points. It doesn't link them all in. At least, it didn't when I last looked, but I think that is ok.

**Comment by not-napoleon** [comment]
@nik9000 am I missing something?  This **is** #58769.  Do you mean that it should close #55552 ?

**Comment by nik9000** [comment]
> @nik9000 am I missing something? This **is** #58769. Do you mean that it should close #55552 ?

Yes! Sorry.

**Review by nik9000 — APPROVED** [review]
LGTM

**Pull request merged**